Cygwin/X: Window placement refinements

Ensure window actually ends up somewhere visible if it tries to create 
itself offscreen. Also, get windows which assume the Windows default 
position placed correctly in multiwindow mode.

Signed-off-by: Jon TURNEY <jon.turney@dronecode.org.uk>
---
 hw/xwin/winmultiwindowwindow.c |    13 ++++++++++++-
 1 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/hw/xwin/winmultiwindowwindow.c b/hw/xwin/winmultiwindowwindow.c
index bb558d1..753dc2c 100644
--- a/hw/xwin/winmultiwindowwindow.c
+++ b/hw/xwin/winmultiwindowwindow.c
@@ -527,6 +527,13 @@ winCreateWindowsWindow (WindowPtr pWin)
   iWidth = pWin->drawable.width;
   iHeight = pWin->drawable.height;
 
+  /* ensure window actually ends up somewhere visible */
+  if (iX > GetSystemMetrics (SM_CXVIRTUALSCREEN))
+    iX = CW_USEDEFAULT;
+
+  if (iY > GetSystemMetrics (SM_CYVIRTUALSCREEN))
+    iY = CW_USEDEFAULT;
+
     if (winMultiWindowGetTransientFor (pWin, &pDaddy))
     {
       if (pDaddy)
@@ -580,7 +580,11 @@ winCreateWindowsWindow (WindowPtr pWin)
   winSelectIcons(pWin, &hIcon, &hIconSmall);
   if (hIcon) SendMessage (hWnd, WM_SETICON, ICON_BIG, (LPARAM) hIcon);
   if (hIconSmall) SendMessage (hWnd, WM_SETICON, ICON_SMALL, (LPARAM) hIconSmall);
- 
+
+  /* If we asked the native WM to place the window, synchronize the X window position */
+  if ((iX == CW_USEDEFAULT) || (iY == CW_USEDEFAULT))
+    winAdjustXWindow(pWin, hWnd);
+
   /* Change style back to popup, already placed... */
   SetWindowLongPtr(hWnd, GWL_STYLE, WS_POPUP | WS_CLIPCHILDREN | WS_CLIPSIBLINGS);
   SetWindowPos (hWnd, 0, 0, 0, 0, 0,

-- 
1.6.4.2

